### ISSUE TITLE
Get consent before tracking users

### ISSUE THREAD

**Issue by ghgr** [hudson-and-thames/mlfinlab#499]
I discovered in a cursory security check that mlfinlab is uploading Personal Identifiable Information to segment.io. It turns out it is described in their [documentation](https://mlfinlab.readthedocs.io/en/latest/additional_information/analytics.html), but at the very bottom and hard to find unless you are explicitly looking for it.

The analytics framework performs additional requests to ipinfo.io and segment.io (first to extract geolocation from your IP, the second to store your data). Additionally, according to the mlfinlab/utils/devadarsh.py file, also your MAC address:

```python
def get_user():
    user = getmac.get_mac_address()
    if user is None:
        user = 'Alfred Borden'
    return user
```

I kindly request that you ask for informed consent prior to collecting this information and offer a way to access, rectify and delete our data. Also, a way to revoke such consent.

**Comment by ghgr** [comment]
Dear Mr Joubert,

Do I understand correctly that: 
 

1. The collection of non-anonymous, non opt-out analytics is described in a link at the bottom of https://mlfinlab.readthedocs.io/en/latest (which as of today it doesn't mention the MAC address) and in Section 11 of the License (of a total of 13 Sections).

2. There's no mention to any of the above points (nor the data collection itself) when installing the package with the recommended method (pip install mlfinlab). I'd expect something like [this](https://stackoverflow.com/questions/24215246/install-using-pip-hangs-on-package-license-agreement).

3. You have no intention of getting explicit consent from the user regarding the upload of their Personally identifiable information (PII) to third party services, including ipinfo.io and segment.io, including of your users in the European Union.

NOTE: (IP Address + MAC Address + Country + Region + City is definitely Personal Identifiable Information)

**Comment by Jackal08** [comment]
Hi @ghgr, 

Thank you for your suggestions. We are busy updating and making changes to improve many of the points listed above.

Will respond with an update today.

**Comment by Jackal08** [comment]
Hi @ghgr, 

Thank you for the suggestions, the following has changed: 

1. Legal docs have been moved to the very top of the [documentation](https://mlfinlab.readthedocs.io/en/latest/).
2. We have triple-checked and can confirm that all of the data we track, including the MAC address is not considered personal information. (Unless we can pair it with additional information that we collect to identify a natural person, which we don't and can't. Thus not personal information.)
3. We have added a [Privacy Policy](https://hudsonthames.org/wp-content/uploads/2021/06/PrivacyPolicy.pdf) and a [GDPR Policy](https://hudsonthames.org/wp-content/uploads/2021/06/GDPR-Policy.pdf).
4. We have updated the [License](https://mlfinlab.readthedocs.io/en/latest/additional_information/license.html#consent-and-the-protection-of-personal-information) to highlight the specific metrics we track.

Regarding the consent on install, this can't be done. We put 3 engineers on this problem and pip doesn't allow custom bash scripts or python files to be executed via the pip install command. Our conclusion is: "it looks like pip redirects all stdout and stderr of the “python setup.py install” command through a special log filter, which reads stdout line by line."

It is for this reason that we can't add a user opt-in on pip install. The alternative is to force the user to declare an environment variable on import of the library which is going to be a pretty disruptive and complicated process for non-technical users. Thus we have decided against an explicit opt-in. The licence agreement, privacy policy, and GDPR policy act as ticket contracts and users consent automatically on typing `pip install mlfinlab`.

Please let me know if there is anything we left out and thank you for helping us to improve the customer experience. 

Best regards
Jacques Joubert

**Comment by ghgr** [comment]
Dear Mr Joubert,

If you claim that the data you are collecting as above described (and also in your website at https://mlfinlab.readthedocs.io/en/latest/additional_information/analytics.html) is completely anonymous (and not pseudonymous, although you have a field that you call UserIDs with the MAC address) and thus not covered by the GDPR, please confirm it by closing this thread.

Otherwise, you are still not getting "informed", "unambiguous" and "specific" consent from the user prior to sharing their PII with third parties. I point you to your own GDPR policy reachable at https://hudsonthames.org/wp-content/uploads/2021/06/GDPR-Policy.pdf , section 3.1.4 Consent:

> - the consent is in the form of an active indication from the Data Subject, i.e., the consent checkbox must not be pre-ticked for the user.

Article 4(11) of the GDPR defines consent as (https://gdpr.eu/article-4-definitions/).

> ‘consent’ of the data subject means any freely given, specific, informed and unambiguous indication of the data subject’s wishes by which he or she, by a statement or by a clear affirmative action, signifies agreement to the processing of personal data relating to him or her;

If you can't get consent the usual mantra is that you just drop the data collection.

I understand your following claims as that you are aware that (a) you are collecting user data without the user's "informed", "unambiguous" and "specific" consent and (b) you have no intention of regularizing this situation:

> - It is for this reason that we can't add a user opt-in on pip install.
> - Thus we have decided against an explicit opt-in.
> - The licence agreement, privacy policy, and GDPR policy act as ticket contracts and users consent automatically on typing pip install mlfinlab.

Please note that just getting explicit consent does not mean being compliant with GDPR, since the user is not "[...] free to give one’s consent or to revoke it." (Section 3.1.4 of your GDPR policy), given that they are forced to accept non-essential tracking not in their legitimate interest. From https://gdpr.eu/Recital-42-Burden-of-proof-and-requirements-for-consent/

> "Consent should not be regarded as freely given if the data subject has no genuine or free choice or is unable to refuse or withdraw consent without detriment."


The following workflow would be in my view acceptable (not legal advice of course):
```
(Each time mlfinlab is imported):
    - If file ~/.config/mlfinlab/tos_accepted does not exist:
        - Display the License with the option to accept (a) or reject (R, default)
        - If not accepted: break
        - If accepted:
            - Save in ~/.config/mlfinlab/tos_accepted
            - Display the Privacy Policy with the option to accept (a) or reject (R, default)
            - If not accepted: break
            - If accepted:
                - Save in ~/.config/mlfinlab/pp_accepted
                - Ask the user if they want to share their usage data, MAC Address and geolocation information with you to improve your services. Inform them that this data sharing is optional and their answer won't change the usability of the software. Option to accept (a) or reject (R, default)
                - If accepted:
                    - Save in ~/.config/mlfinlab/data_sharing_accepted

Finally, any time your analytics software intends to upload check if the file data_sharing_accepted exists. If it doesn't, just don't upload the data.
```

**Comment by Jackal08** [comment]
Hi @ghgr, 

I am quite confident that the analytics we track in the libraries are all above board. They don't need to be anonymous, they only need to be in a format that they don't qualify as PII which means that we can't identify natural persons, they are also of a nature that we can't identify legal persons either. 

The GDPR policy that you are referencing has been written to include the PII data that we do collect from users who sign up for marketing materials such as our Ebooks. It's that data that we get explicit consent for. 

The purpose of the analytics in libraries is largely there to force large institutions to comply with our license and to incentivize them to take up an enterprise license that does not track analytics. This is because their compliance department will never let software with analytics run in-house. 

For everyone else who isn't a large institution and is more flexible w.r.t their compliance needs - can use the best tech without the big price tag for an enterprise license. It's for this reason that the analytics we track will never be optional. We are specifically saying that the use of the libraries without an enterprise license comes with analytics and users consent to that when they pip install.

You have a good idea with the config file and we have added a ticket to our project board to investigate if it can be used inside a python library. The idea is that upon the first import of the library we can explicitly inform the user that we collect data and they can agree or not. If they agree then the library will be imported, else they can't use the tools.

My hope is that the config idea works since it would be an elegant solution. This could be added to our next release. 

I will leave this ticket open for 24 hours before closing it, in case you would like to expand. 

Thank you - your feedback has indeed made our libraries a better product.

**Comment by ghgr** [comment]
Dear Mr Joubert,

I appreciate the modifications you made to the libraries and applaud your endeavour of implementing a robust ML financial library such as this one.

I insist that the data you collect is definitely considered PII according to the GDPR. I cite [Recital 30 | Online identifiers for profiling and identification](https://gdpr.eu/https://gdpr.eu/recital-30-online-identifiers-for-profiling-and-identification/):

> Natural persons may be associated with online identifiers provided by their devices, applications, tools and protocols, such as internet protocol addresses, cookie identifiers or other identifiers such as radio frequency identification tags. This may leave traces which, in particular when combined with unique identifiers and other information received by the servers, may be used to create profiles of the natural persons and identify them.

IP Addresses are explicitly mentioned, which are arguably less consistent that MAC Adresses (and which are still collected when making the HTTP request). The reason for this is that you can build a profile of my MAC activities (pseudonym), and if in the future you manage to map this MAC to my name, you'll automatically have a profile of a natural person.

I believe this thread summarizes our points of view so if you don't have anything else to add you can close it.

**Comment by rspadim** [comment]
hi guys i was reading this topic

maybe a workaround is implement some functions at lib at if you don't call a "auth()" you have limited access, and to allow this auth() function you should accept a license at hudson and thames that agree with the tracking of your uses and help ht to know what functions are more used / develop something with more knowledge about users

this helps?

**Comment by Jackal08** [comment]
@ghgr, 

Thank you for highlighting this. I am comfortable that we have put the systems in place to protect us from GDPR compliance issues. We will investigate the use of config files and if it works, we will certainly add it. We do think that explicit confirmation of the analytics would be a good idea but as it stands right now - there isn't a possible solution to this as PIP stops it.

(The easiest solution is to remove public access altogether and then when users purchase libraries from the company website, we can have explicit opt-in and a lot more flexibility.)

On the risk of our definition of PII is wrong, we would then have to pursue further legal action. We are taking the stance from multiple sources that IP Addresses and MAC addresses are not PII so long as they cant be combined with additional data sources to identify a natural person. We have no such additional data and there is no way for us to infer this relationship. 

I'd like to highlight that the data we track is very similar in nature to Google analytics - which doesn't use cookies. Users don't explicitly opt into those analytics either, and they are on almost every single website you visit. 

@rspadim 

The truth is that keeping MlFinLab and PortfolioLab publicly available is causing more problems for us as a company, from both a compliance and additional engineering overhead. It cost us a great deal of money to keep that library publicly available, and the marginal benefit is getting smaller and smaller. 

As a company, we are leaning towards removing its public access altogether and focusing on SAAS enterprise sales exclusively. However, I hope that this day never comes and we can continue to provide what is a huge benefit to the community, for free (non-commercial use).

**Note:**

* We have no intention of removing public access. We continue to fight to keep it publicly available. 
* I am just highlighting the major pain in the ass - maintaining public access has caused.
* We continue to build tools that we hope the community will love.

**Comment by ghgr** [comment]
Thank you for taking this matter seriously.

After your declarations about the problems caused by keeping it public I feel obliged to offer a counterpoint, in case bystanders are following this thread.

As a developer myself, I've dealt with the intricacies of the GDPR, and I agree that sometimes it can be -as you succinctly put- a pain in the ass. Nevertheless, this is a law that has a very good reason to exist. We are living new times and the risks of the unregulated data collection and processing with technical means escapes our collective understanding. We are usually taught by our parents that we should not eat spoiled food, or drive under the influence of alcohol. But the risks of aggregated data are at least as large, clearly illustrated by the [scandal of the Cambridge Analytica](https://en.wikipedia.org/wiki/Facebook%E2%80%93Cambridge_Analytica_data_scandal).

That's why the GDPR exists. To force companies dealing with data to have a basic hygiene when processing it. It has an inclusive definition of what Personally Identifiable Information (PII) means, to avoid loopholes of re-anonymization (as illustrated above with the pseudonym profile based on the MAC address), and has draconian fines to force companies to comply. It has been carefully designed to avoid common loopholes like burying the information under a wall of legalese or coercing the user to accept. Or even offering the "accept" checkbox already checked, so they unwillingly comply.

Nevertheless we, as adults, can offer our personal data, but only if they inform us clearly (which basically boils down to doing it in good-faith), and if we are willing (again, no coercion) to do so.

You are free to stop offering a free tier altogether, but I kindly ask you not to pin it on the law. It's if a restaurant owner says that the health codes are too strict and thus it's their fault that they won't open. All that GDPR is asking is that you inform the user and ask for their permission, without coercion. If you'd rather stop offering the product before doing that (which is not my intention at all) that's your strategic decision.

Tackling some of your points:

- Websites using Google Analytics definitely need to ask for permission (the infamous cookie banner), which is debatable how efficient it is informing users.

- You claim that "the easiest solution is to remove public access altogether". I disagree. The easiest solution is to just stop collecting PII from the user, which is not necessary anyway.

In summary, the GDPR does not force you to remove the free tier. Again, all it requires from you is to ask for free, informed, specific and unambiguous consent prior to uploading your user's PII. That, or don't upload their PII.

As a final remark, I agree with your last sentence. I'm sure this is a library that the community loves.

**Comment by vasili-zolotov** [comment]
@Jackal08 can you elaborate on the bullet "Function calls" mentioned in your documentation? Are you logging just method names being called or are you logging additional details (for example function arguments).

**Comment by Jackal08** [comment]
Hi @vasili-zolotov, 

In the context: We only track function names, no additional information leaves your PC other than those mentioned. It's the same web analytics found in the vast majority of SAAS products.

We do offer an enterprise license for companies that want to have the analytics removed.

**Comment by jeffneuen** [comment]
@vasili-zolotov you can see for yourself if you change the loglevel to debugging in your own code while using the library. It shows the request being to used to push out the data.

**Comment by Jackal08** [comment]
Hi All, 

The ICO launched an investigation on the basis that we were tracking personal information. They have, after a thorough study, ruled that we do in fact, NOT track personal information and everything is above board. 

Wanted to put everyone's mind at ease. 

Best regards
Jacques Joubert
